apps/glusterfs/volume_entry_allocate.go

```diff
@@ -205,14 +206,46 @@ func (v *VolumeEntry) replaceBrickInVolume(db *bolt.DB, executor executors.Execu
 		}
 
 		// Try to allocate a brick on this device
-		newBrickEntry := newDeviceEntry.NewBrickEntry(oldBrickEntry.Info.Size,
-			float64(v.Info.Snapshot.Factor),
-			v.gidRequested, v.Info.Id)
+		// NewBrickEntry would deduct storage from device entry
+		// which we will save to disk, hence reload the latest device
+		// entry to get latest storage state of device
+		err = db.Update(func(tx *bolt.Tx) error {
+			newDeviceEntry, err := NewDeviceEntryFromId(tx, deviceId)
+			if err != nil {
+				return err
+			}
+			newBrickEntry = newDeviceEntry.NewBrickEntry(oldBrickEntry.Info.Size,
+				float64(v.Info.Snapshot.Factor),
+				v.gidRequested, v.Info.Id)
+			err = newDeviceEntry.Save(tx)
+			if err != nil {
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
 
 		// Determine if it was successful
 		if newBrickEntry == nil {
 			continue
 		}
+
+		defer func() {
+			if e != nil {
+				db.Update(func(tx *bolt.Tx) error {
+					newDeviceEntry, err = NewDeviceEntryFromId(tx, newBrickEntry.Info.DeviceId)
+					if err != nil {
+						return err
+					}
+					newDeviceEntry.StorageFree(newBrickEntry.TotalSize())
+					newDeviceEntry.Save(tx)
```

Comment:
I am not entirely sure about this. I guess it might be ok in this case. Also, we are not fully consistent with this across the code base. Inclined to merge it like this and we can sort it out later in full generality..